### DISCUSSION TITLE
verbatim element in the LaTeX namespace

### DISCUSSION THREAD

**Discussion by FrankMittelbach** [latex3/tagging-project#7] [LaTeX namespace]
_General disclaimer:_ all of the tag names and often also the underlying structures are still subject to discussion and are not meant to be final.

#### Verbatim blocks

There are 6 tags in the LaTeX namespace related to _verbatim_ environment (or closely related environments), i.e., to a line oriented display block in which the concept of a _line_ is semantically relevant and lines can have labels (typically line numbers for reference) and comments related to the line.

The tags are:

- `<verbatim>` enclosing the overall structure
- `<verbatim-line>` enclosing each _line_ of the structure
- `<verbatim-text>` enclosing the text in a _line_ (the main data)
- `<verbatim-label>` enclosing the label in a _line_ if present
- `<verbatim-comment>` enclosing the comment text in a _line_ if present
- `<verbatim-wrapping-indicator>` enclosing the material used to separate line parts, if a _line_ for reasons of physical limitations has to be split across 2 or more physical lines in the display
 

```
 {verbatim} ::=  <verbatim> ( {verbatim line} )*  </verbatim>

 {verbatim line} ::=  <verbatim-line>
                       {normal verbatim line} | {wrapped verbatim line}
                      </verbatim-line>

{normal verbatim line}  ::= [ {verbatim label} ] {verbatim text} [ {verbatim comment} ]
{wrapped verbatim line} ::= {normal verbatim line}
                            ( [ {verbatim wrapping indicator} ]
                                {normal verbatim line}          )+

{verbatim wrapping indicator} ::= <verbatim-wrapping-indicator> {mc chunk}  </verbatim-wrapping-indicator>


{verbatim label}   ::=  <verbatim-label>   {label elements} </verbatim-label>
{verbatim text}    ::=  <verbatim-text>    {mc chunk}       </verbatim-text>
{verbatim comment} ::=  <verbatim-comment> {mc chunk}       </verbatim-comment>

```

##### Notes

- `{verbatim text}` can denote an empty _verbatim line_ (if its mc chunk is empty) but it has to be there (within code empty lines typically have semantic meaning).

- If there is not enough space _verbatim lines_ can get wrapped for printing, typically with a "wrapping indicator", such as some sort of arrow. In that case, the "wrapped lines" are considered a single `{verbatim line}` structure and the individual parts are separated by a structural element (that is typically, but not necessarily, roll-mapped to Artifact)

- Lamport-LaTeX only understands a simple _verbatim_ without labels and comments or line wrapping but several extension packages support such substructures.

##### Role-mapping

tbd

**Comment by FrankMittelbach** [comment]
###### Current LaTeX rolemapping for 1.7 (2023-08-09):

```
<verbatim> ::= P
<verbatim-line> ::= Span
<verbatim-label> ::= Lbl
<verbatim-text> ::= Code
<verbatim-comment> ::= Span
<verbatim-wrapping-indicator> ::= Artifact      (by default, or Span when it is relevant)
```

Notes:

- Should that be `<verbatim-line> ::= Span placement:block`?

**Comment by FrankMittelbach** [comment]
###### Suggestion by Matthew for role mapping (2023-08-09):

```
<verbatim> ::= Code
<verbatim-line> ::= Span placement:block
<verbatim-label> ::= Lbl
<verbatim-text> ::= Span placement:inline
<verbatim-comment> ::= Span placement:inline
<verbatim-wrapping-indicator> ::= Span placement:inline
```

**Reply by car222222** [reply]
In fact, @FrankMittelbach did in his introduction to this element, show clearly that the 'verbatim nature' of these elements is not, in fact, their important feature: rather, they are merely things that have a substructure of "lines".  Thus "poem" should also be included here:-).  

He wrote: 
" i.e., . . . a line oriented display block in which the concept of a line is semantically relevant "

**Reply by FrankMittelbach** [reply]
@car222222 oh common Chris, it doesn't clearly show that it is not important, it only shows that I may not be able to express myself perfectly. Read it as as  "a line oriented display block of predominately verbatim/code material and in which the concept of a line is semantically relevant."  Better? Clearer?

The distinction I wanted to make is  that in comparison to  "inline code" inside a paragraph of text where line breaks are merely the result of j&h and are of no semantic significance.

As to "poems":  they would have a similar structure I guess but their own set of tags imho.

**Comment by LupinoTech** [comment]
I don't quite understand why `verbatim-comment` is necessary. Maybe, i'm too focused on source code as verbatim, but to me, comments belong to the verbatim content and if you start tagging them seperately, why not also tag keywords, methods, class names, etc.? One could argue that comments that explain the rest of the verbatim material on a meta-ish level do not belong inside the verbatim environment, in the first place.

**Reply by car222222** [reply]
Yes, this is a very good analysis by @LupinoTech. 

One (extremely major) problem here is the use of the term "verbatim" as the element name; it is extremely misleading 
(and is probably not even a semantic concept!)

This should definitely be changed (right now), to something 
like "code-listing" or "code-block", . . .

**Reply by FrankMittelbach** [reply]
@LupinoTech the intention here is not to substructure the code material that would indeed ask for various other elements such as class, keyword, method, etc., the intention is to separate out the 2  optional items per line that are not "code" but appear in  certain documents as part of each or some line in such a display. Take for example, `<verbatim-label>`: in some languages, say Basic, line numbers are are in fact part of the code and in that case they (depending on the  author's intentions)  are best  tagged as being part of the code proper andd not as a label in the sense of  `<verbatim-label>`. On the other hand, if the author line numbers the display to be able to reference the code lines  in outer text, then these line number should be marked as `<verbatim-label>` because they are not part of the code in the line.

With the same reasoning there is `<verbatim-comment>`. It is meant for the case that there are comments related to the code in the line and placed next to it but aren't part of the code in that line itself. Again there are two cases: one is that the code line contains a comment using the syntax of the language in which it is written, e.g. `/* ...*/` etc. in which case it would be perfectly proper to just consider it part of the code and keep it inside `<verbatim-text>`  or the comment could have been presented in a way that it is technically not code in that language but distinguished by other means (e.g., font choice) so that for the reader as a consumer the distinction is clear. But to lift out just the code from the display it requires therefore to be marked as separate from the code proper on that line.

**Reply by car222222** [reply]
I think that for the first case the comment should be treated as 
just part of the code, with no special tagging.  This is best partly because there are so many different syntaxes and 
placements for such comments, even within one language! --

"the code line contains a comment using the syntax of the language in which it is written, e.g. /* ...*/ etc."

So anyways, it should be "a syntax"!

**Reply by FrankMittelbach** [reply]
Ok perhaps again a mother tongue issue. That is what I meant.  If it is part of the code as such then putting it there is proper (aka the right choice), but there is the other form in which case it would make the code invalid if thought as part of the code. Maybe I should have made 2 bullet points rather than a longish sentence to make the distinction between the two cases clear.

**Reply by FrankMittelbach** [reply]
@car222222 I don't really understand why you think that `verbatim` is misleading (or even extremely misleading) or why therefore `code-listing` or `code-block` is really better. 

As I said at the very top "names" are really still  subject to change so if we decide on others fine, right now all I'm saying I don't see the "why" here.
However, whatever we decide, I think it is very helpful if as a convention if the top-level name (which currently is `verbatim`) appears in sub-level names, e.g., `verbatim-line`, etc.    That in turn would then mean you end up with relatively long names which is not a big issue but something to keep in mind.

But back the name: `verbatim`seems to me (at least in the TeX world, but not just there imo)  a fairly clear and understandable concept and  in my opinion is capturing the situation better than "code" which  has to me always the feel of "computer programming code" at least as a first assocication. I know that in English "code" has also quite a number of other meanings  but something like "code of honor" isn't meant here either (even though you could "present it verbatim").

But computer code isn't all that you would put into a verbatim environment, which is why verbatim in the sense of the German word "wörtlich" conveys, to me at least, the right meaning.

**Reply by LupinoTech** [reply]
I wrote "I don't quite understand"... now, i do and I get why the idea of `verbatin-comment` might be necessary: one use-case could be to indicate whitespaces, where `␣` would be tagged with that thing, right? 

The only aspect i'd change is the naming, since "comment" seems to be too abmiguous in this context. How about `verbatim-meta` or `verbatim-context`?

**Comment by u-fischer** [comment]
> it could be presented in a way that it is technically not code in that language but distriguished by other means (e.g., font choice)

Thinking about it, the `verbatim-wrapping-indicator` is actually a comment of this type. Wouldn't it then make sense to drop the special wrapping case and instead simply allow that `verbatim-text` and `verbatim-comment`  can appear more than once in a `verbatim-line`? That would than also cover cases, where pieces of code are commented with arrows and external text boxes (it wouldn't be trivial to tag, but at least the formal description would allow to handle it).

**Reply by FrankMittelbach** [reply]
I like to defend the current formal syntax in that respect.  The `<verbatim-wrapping-indicator>` is an important  element in such a display but it is typically marked up differently (i.e., normally as an Artifact). The current formal syntax already supports more than one `<verbatim-text>` etc per `<verbatim-line>`, but only if separated by `<verbatim-wrapping-indicator>`, which I think is not unreasonable and offers consistent but flexible tagging.

**Reply by u-fischer** [reply]
> but it is typically marked up differently (i.e., normally as an Artifact)

Ah, right, I forgot about artifact. Then we perhaps we need two types of "comments" here?

And regarding more than one `verbatim-text`: I was wondering about annotations like in this answer: https://tex.stackexchange.com/a/223553/2388. I'm not sure if one should use alternating text/comments chunks, or if one should collect the annotations at the end of the line or even the verbatim block and use /Ref, but one clearly has more than one comment for a line.
